File: frontend/src/lib/pof/codec.ts
// The .pof wire format, as specified in /docs/format:
//
//   magic     "POF1"                 4 bytes
//   version   u16 little-endian      2 bytes
//   body      postcard(Body)         variable
//   checksum  blake2b-256(body)      32 bytes
//
// Travels as base64url wherever it is text. This is a TypeScript mirror used for display and
// parsing; pof-verify (Rust, compiled to WASM) is the reference and the only thing that
// decides a verdict. `scripts/verify-fixtures.ts` checks this mirror byte-for-byte against the
// proofs the Rust prover wrote to fixtures/.

import type { Claim, Envelope } from '../data/types.ts'
import { blake2b } from './blake2b.ts'
import { concat, equal, fromHex, hex } from './bytes.ts'

export const MAGIC = new Uint8Array([0x50, 0x4f, 0x46, 0x31]) // "POF1"
export const FORMAT_VERSION = 1
/** 1 unit = 0.125 ZEC. Threshold claims are proven in whole units. */
export const ZAT_PER_UNIT = 12_500_000n
/** Total supply bound, in zatoshi (pof_core::MAX_ZATOSHI). */
export const MAX_ZATOSHI = 2_100_000_000_000_000
const MAX_U32 = 0xffffffff

const CLAIM_TAG: Record<Claim['kind'], number> = {
  HoldsAtLeast: 0,
  HoldsExactly: 1,
  ReceivedPayment: 2,
  ReceivedAtLeastSince: 3,
}

// ── writer ────────────────────────────────────────────────────────────────

class Writer {
  parts: number[] = []
  varint(n: number) {
    if (!Number.isSafeInteger(n) || n < 0) throw new RangeError(`varint out of range: ${n}`)
    do {
      let byte = n % 128
      n = Math.floor(n / 128)
      if (n > 0) byte |= 0x80
      this.parts.push(byte)
    } while (n > 0)
  }
  fixed(b: Uint8Array) {
    for (const x of b) this.parts.push(x)
  }
  bytes(b: Uint8Array) {
    this.varint(b.length)
    this.fixed(b)
  }
  done() {
    return new Uint8Array(this.parts)
  }
}

function writeHead(w: Writer, e: Envelope) {
  const c = e.claim
  w.varint(CLAIM_TAG[c.kind])
  switch (c.kind) {
    case 'HoldsAtLeast':
    case 'HoldsExactly':
      w.varint(c.zatoshi)
      break
    case 'ReceivedPayment':
      w.fixed(fromHex(c.txid))
      w.varint(c.zatoshi)
      break
    case 'ReceivedAtLeastSince':
      w.varint(c.zatoshi)
      w.varint(c.fromHeight)
      break
  }
  w.fixed(fromHex(e.audience))
  w.fixed(fromHex(e.binding))
  w.varint(e.anchor.height)
  w.fixed(fromHex(e.anchor.ncRoot))
  w.fixed(fromHex(e.anchor.nfRoot))
  w.varint(e.issuedAt)
  w.varint(e.expiresAt)
  w.fixed(fromHex(e.revocation))
}

export function encodeBody(e: Envelope): Uint8Array {
  const w = new Writer()
  writeHead(w, e)
  w.varint(e.evidence.publicInputs.length)
  for (const pi of e.evidence.publicInputs) w.fixed(fromHex(pi))
  w.bytes(e.evidence.proof)
  w.fixed(fromHex(e.evidence.signature))
  return w.done()
}

export function encode(e: Envelope): Uint8Array {
  const body = encodeBody(e)
  const version = new Uint8Array([FORMAT_VERSION & 0xff, FORMAT_VERSION >> 8])
  return concat(MAGIC, version, body, blake2b(body))
}

/** Re-seal a body with a fresh checksum — what a forger would do after editing it. */
export function reseal(file: Uint8Array): Uint8Array {
  const body = file.subarray(6, file.length - 32)
  return concat(file.subarray(0, 6), body, blake2b(body))
}

// ── reader ────────────────────────────────────────────────────────────────

class Reader {
  o = 0
  private b: Uint8Array
  constructor(b: Uint8Array) {
    this.b = b
  }
  /** Also the timestamp bound: pof_core refuses issue and expiry times above 2^53 − 1. */
  varint(): number {
    let n = 0
    let mul = 1
    for (let i = 0; i < 8; i++) {
      const byte = this.u8()
      n += (byte & 0x7f) * mul
      if (!(byte & 0x80)) {
        // A zero last group pads a shorter encoding of the same value; decoding stays injective.
        if (byte === 0 && i > 0) throw new Error('non-canonical varint')
        if (!Number.isSafeInteger(n)) throw new Error('value exceeds 2^53')
        return n
      }
      mul *= 128
    }
    throw new Error('varint too long')
  }
  u32(): number {
    const n = this.varint()
    if (n > MAX_U32) throw new Error('value overflows u32')
    return n
  }
  amount(): number {
    const n = this.varint()
    if (n > MAX_ZATOSHI) throw new Error('amount exceeds the 21M ZEC supply')
    return n
  }
  u8(): number {
    if (this.o >= this.b.length) throw new Error('unexpected end of body')
    return this.b[this.o++]!
  }
  fixed(n: number): Uint8Array {
    if (this.o + n > this.b.length) throw new Error('unexpected end of body')
    const out = this.b.slice(this.o, this.o + n)
    this.o += n
    return out
  }
  bytes(max: number): Uint8Array {
    const n = this.varint()
    if (n > max) throw new Error(`field of ${n} bytes exceeds the ${max}-byte limit`)
    return this.fixed(n)
  }
  get rest() {
    return this.b.length - this.o
  }
}

export type DecodeResult =
  | { ok: true; envelope: Envelope; checksum: string }
  | { ok: false; reason: string }

export function decode(file: Uint8Array): DecodeResult {
  if (file.length < 4 + 2 + 32 + 1) return { ok: false, reason: 'Too short to be a proof. The paste may be truncated.' }
  if (!equal(file.subarray(0, 4), MAGIC)) return { ok: false, reason: 'Not a proof file: the POF1 magic bytes are missing.' }
  const version = file[4]! | (file[5]! << 8)
  if (version !== FORMAT_VERSION) return { ok: false, reason: `Unknown format version ${version}. This verifier reads version ${FORMAT_VERSION}.` }

  const body = file.subarray(6, file.length - 32)
  const checksum = file.subarray(file.length - 32)
  if (!equal(blake2b(body), checksum)) {
    return { ok: false, reason: 'Checksum mismatch. The file was truncated or damaged in transit — this is not the same as a forgery.' }
  }

  try {
    const r = new Reader(body)
    const tag = r.varint()
    let claim: Claim
    switch (tag) {
      case 0: claim = { kind: 'HoldsAtLeast', zatoshi: r.amount() }; break
      case 1: claim = { kind: 'HoldsExactly', zatoshi: r.amount() }; break
      case 2: { const txid = hex(r.fixed(32)); claim = { kind: 'ReceivedPayment', txid, zatoshi: r.amount() }; break }
      case 3: { const zatoshi = r.amount(); claim = { kind: 'ReceivedAtLeastSince', zatoshi, fromHeight: r.u32() }; break }
      default: throw new Error(`unknown claim tag ${tag}`)
    }
    const audience = hex(r.fixed(32))
    const binding = hex(r.fixed(32))
    const height = r.u32()
    const ncRoot = hex(r.fixed(32))
    const nfRoot = hex(r.fixed(32))
    const issuedAt = r.varint()
    const expiresAt = r.varint()
    if (issuedAt > expiresAt) throw new Error('issued after it expires')
    const revocation = hex(r.fixed(16))
    const n = r.varint()
    if (n > 16) throw new Error('too many public inputs')
    const publicInputs: string[] = []
    for (let i = 0; i < n; i++) publicInputs.push(hex(r.fixed(32)))
    const proof = r.bytes(16 * 1024)
    const signature = hex(r.fixed(64))
    if (r.rest !== 0) throw new Error('trailing bytes after the envelope')
    return {
      ok: true,
      envelope: {
        version,
        claim,
        audience,
        binding,
        anchor: { height, ncRoot, nfRoot },
        issuedAt,
        expiresAt,
        revocation,
        evidence: { publicInputs, proof, signature },
      },
      checksum: hex(checksum),
    }
  } catch (err) {
    return { ok: false, reason: `Envelope does not parse: ${(err as Error).message}.` }
  }
}

/** True when `binding` is the all-zero "unbound" value. */
export const isUnbound = (binding: string) => /^0{64}$/.test(binding)

File: backend/crates/pof-core/src/codec.rs
//! Byte-exact encoder and decoder. The encoding is postcard's: unsigned integers as LEB128
//! varints, fixed arrays as raw bytes, vectors as a varint length then elements, enums as a
//! varint tag then fields. Written by hand so the TypeScript mirror (frontend
//! `src/lib/pof/codec.ts`) and this crate can be checked against the same golden vectors.

use crate::{
    hash::blake2b_256, Anchor, Claim, Envelope, Evidence, FORMAT_VERSION, MAGIC_BYTES, MAX_PROOF_BYTES, MAX_PUBLIC_INPUTS,
    MAX_TIMESTAMP, MAX_ZATOSHI,
};

/// Why a file failed to parse. Each variant is a different message in the UI.
#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
pub enum DecodeError {
    #[error("Too short to be a proof. The paste may be truncated.")]
    TooShort,
    #[error("Not a proof file: the POF1 magic bytes are missing.")]
    BadMagic,
    #[error("Unknown format version {0}. This verifier reads version 1.")]
    UnknownVersion(u16),
    #[error("Checksum mismatch. The file was truncated or damaged in transit — this is not the same as a forgery.")]
    Checksum,
    #[error("Envelope does not parse: {0}.")]
    Body(String),
    #[error("Input is not base64url text or a .pof file.")]
    NotBase64,
}

// ── writer ────────────────────────────────────────────────────────────────

struct Writer(Vec<u8>);

impl Writer {
    fn varint(&mut self, mut n: u64) {
        loop {
            let byte = (n & 0x7f) as u8;
            n >>= 7;
            if n == 0 {
                self.0.push(byte);
                return;
            }
            self.0.push(byte | 0x80);
        }
    }
    fn fixed(&mut self, b: &[u8]) {
        self.0.extend_from_slice(b);
    }
    fn bytes(&mut self, b: &[u8]) {
        self.varint(b.len() as u64);
        self.fixed(b);
    }
}

fn write_head(w: &mut Writer, e: &Envelope) {
    w.varint(e.claim.tag() as u64);
    match &e.claim {
        Claim::HoldsAtLeast { zatoshi } | Claim::HoldsExactly { zatoshi } => w.varint(*zatoshi),
        Claim::ReceivedPayment { txid, zatoshi } => {
            w.fixed(txid);
            w.varint(*zatoshi);
        }
        Claim::ReceivedAtLeastSince { zatoshi, from_height } => {
            w.varint(*zatoshi);
            w.varint(*from_height as u64);
        }
    }
    w.fixed(&e.audience);
    w.fixed(&e.binding);
    w.varint(e.anchor.height as u64);
    w.fixed(&e.anchor.nc_root);
    w.fixed(&e.anchor.nf_root);
    w.varint(e.issued_at);
    w.varint(e.expires_at);
    w.fixed(&e.revocation);
}

/// The statement: every body field except the evidence. This is what the proof binds to.
pub(crate) fn encode_head(e: &Envelope) -> Vec<u8> {
    let mut w = Writer(Vec::with_capacity(160));
    write_head(&mut w, e);
    w.0
}

fn encode_body(e: &Envelope) -> Vec<u8> {
    let mut w = Writer(Vec::with_capacity(160 + e.evidence.proof.len() + 32 * e.evidence.public_inputs.len()));
    write_head(&mut w, e);
    w.varint(e.evidence.public_inputs.len() as u64);
    for pi in &e.evidence.public_inputs {
        w.fixed(pi);
    }
    w.bytes(&e.evidence.proof);
    w.fixed(&e.evidence.signature);
    w.0
}

pub fn encode(e: &Envelope) -> Vec<u8> {
    let body = encode_body(e);
    let mut out = Vec::with_capacity(4 + 2 + body.len() + 32);
    out.extend_from_slice(&MAGIC_BYTES);
    out.extend_from_slice(&FORMAT_VERSION.to_le_bytes());
    out.extend_from_slice(&body);
    out.extend_from_slice(&blake2b_256(&body));
    out
}

// ── reader ────────────────────────────────────────────────────────────────

struct Reader<'a> {
    b: &'a [u8],
    o: usize,
}

impl<'a> Reader<'a> {
    fn u8(&mut self) -> Result<u8, String> {
        let x = *self.b.get(self.o).ok_or("unexpected end of body")?;
        self.o += 1;
        Ok(x)
    }
    fn varint(&mut self) -> Result<u64, String> {
        let mut n: u64 = 0;
        for i in 0..10 {
            let byte = self.u8()?;
            let chunk = (byte & 0x7f) as u64;
            if i == 9 && chunk > 1 {
                return Err("varint overflows u64".into());
            }
            n |= chunk << (7 * i);
            if byte & 0x80 == 0 {
                // A zero last group pads a shorter encoding of the same value. Refused so that
                // decoding is injective: one proof, one byte string, one checksum.
                if byte == 0 && i > 0 {
                    return Err("non-canonical varint".into());
                }
                return Ok(n);
            }
        }
        Err("varint too long".into())
    }
    fn u32(&mut self) -> Result<u32, String> {
        u32::try_from(self.varint()?).map_err(|_| "value overflows u32".into())
    }
    fn fixed<const N: usize>(&mut self) -> Result<[u8; N], String> {
        let end = self.o.checked_add(N).filter(|&e| e <= self.b.len()).ok_or("unexpected end of body")?;
        let mut out = [0u8; N];
        out.copy_from_slice(&self.b[self.o..end]);
        self.o = end;
        Ok(out)
    }
    /// A length prefix, bounded in u64 before narrowing: on wasm32 `as usize` would wrap 2^32 + k to k.
    fn len(&mut self, max: usize, too_long: impl FnOnce(u64) -> String) -> Result<usize, String> {
        let n = self.varint()?;
        if n > max as u64 {
            return Err(too_long(n));
        }
        Ok(n as usize)
    }
    fn bytes(&mut self, max: usize) -> Result<Vec<u8>, String> {
        let n = self.len(max, |n| format!("field of {n} bytes exceeds the {max}-byte limit"))?;
        let end = self.o.checked_add(n).filter(|&e| e <= self.b.len()).ok_or("unexpected end of body")?;
        let out = self.b[self.o..end].to_vec();
        self.o = end;
        Ok(out)
    }
}

fn amount(z: u64) -> Result<u64, String> {
    if z > MAX_ZATOSHI {
        Err("amount exceeds the 21M ZEC supply".into())
    } else {
        Ok(z)
    }
}

fn read_body(body: &[u8]) -> Result<Envelope, String> {
    let mut r = Reader { b: body, o: 0 };
    let claim = match r.varint()? {
        0 => Claim::HoldsAtLeast { zatoshi: amount(r.varint()?)? },
        1 => Claim::HoldsExactly { zatoshi: amount(r.varint()?)? },
        2 => {
            let txid = r.fixed::<32>()?;
            Claim::ReceivedPayment { txid, zatoshi: amount(r.varint()?)? }
        }
        3 => {
            let zatoshi = amount(r.varint()?)?;
            Claim::ReceivedAtLeastSince { zatoshi, from_height: r.u32()? }
        }
        t => return Err(format!("unknown claim tag {t}")),
    };
    let audience = r.fixed::<32>()?;
    let binding = r.fixed::<32>()?;
    let anchor = Anchor { height: r.u32()?, nc_root: r.fixed::<32>()?, nf_root: r.fixed::<32>()? };
    let issued_at = r.varint()?;
    let expires_at = r.varint()?;
    if issued_at.max(expires_at) > MAX_TIMESTAMP {
        return Err("timestamp exceeds 2^53 - 1".into());
    }
    if issued_at > expires_at {
        return Err("issued after it expires".into());
    }
    let revocation = r.fixed::<16>()?;
    let n = r.len(MAX_PUBLIC_INPUTS, |_| "too many public inputs".into())?;
    let mut public_inputs = Vec::with_capacity(n);
    for _ in 0..n {
        public_inputs.push(r.fixed::<32>()?);
    }
    let proof = r.bytes(MAX_PROOF_BYTES)?;
    let signature = r.fixed::<64>()?;
    if r.o != body.len() {
        return Err("trailing bytes after the envelope".into());
    }
    Ok(Envelope {
        claim,
        audience,
        binding,
        anchor,
        issued_at,
        expires_at,
        revocation,
        evidence: Evidence { public_inputs, proof, signature },
    })
}

/// Parse a `.pof` file. On success also returns the hex checksum.
pub fn decode(file: &[u8]) -> Result<(Envelope, [u8; 32]), DecodeError> {
    if file.len() < 4 + 2 + 32 + 1 {
        return Err(DecodeError::TooShort);
    }
    if file[..4] != MAGIC_BYTES {
        return Err(DecodeError::BadMagic);
    }
    let version = u16::from_le_bytes([file[4], file[5]]);
    if version != FORMAT_VERSION {
        return Err(DecodeError::UnknownVersion(version));
    }
    let body = &file[6..file.len() - 32];
    let mut checksum = [0u8; 32];
    checksum.copy_from_slice(&file[file.len() - 32..]);
    if blake2b_256(body) != checksum {
        return Err(DecodeError::Checksum);
    }
    let env = read_body(body).map_err(DecodeError::Body)?;
    Ok((env, checksum))
}

// ── base64url (RFC 4648 §5, no padding; standard base64 is accepted on input) ──

const B64: &[u8; 64] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";

pub fn to_base64url(b: &[u8]) -> String {
    let mut s = String::with_capacity(b.len().div_ceil(3) * 4);
    for chunk in b.chunks(3) {
        let n = (chunk[0] as u32) << 16 | (*chunk.get(1).unwrap_or(&0) as u32) << 8 | *chunk.get(2).unwrap_or(&0) as u32;
        s.push(B64[(n >> 18) as usize & 63] as char);
        s.push(B64[(n >> 12) as usize & 63] as char);
        if chunk.len() > 1 {
            s.push(B64[(n >> 6) as usize & 63] as char);
        }
        if chunk.len() > 2 {
            s.push(B64[n as usize & 63] as char);
        }
    }
    s
}

pub fn from_base64url(s: &str) -> Result<Vec<u8>, DecodeError> {
    let mut out = Vec::with_capacity(s.len() * 3 / 4);
    let (mut buf, mut bits) = (0u32, 0u32);
    for ch in s.trim().trim_end_matches('=').bytes() {
        if ch.is_ascii_whitespace() {
            continue;
        }
        let ch = match ch {
            b'+' => b'-',
            b'/' => b'_',
            c => c,
        };
        let v = B64.iter().position(|&c| c == ch).ok_or(DecodeError::NotBase64)? as u32;
        buf = (buf << 6) | v;
        bits += 6;
        if bits >= 8 {
            bits -= 8;
            out.push((buf >> bits) as u8);
        }
    }
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::*;

    pub(crate) fn sample(claim: Claim) -> Envelope {
        Envelope {
            claim,
            audience: audience_hash("pof-credit:usdc-pool-1"),
            binding: [7u8; 32],
            anchor: Anchor { height: 3_491_040, nc_root: [1; 32], nf_root: [2; 32] },
            issued_at: 1_790_035_200,
            expires_at: 1_790_640_000,
            revocation: [9; 16],
            evidence: Evidence { public_inputs: vec![[3; 32]; 9], proof: vec![0xab; 2_208], signature: [5; 64] },
        }
    }

    #[test]
    fn round_trip_every_claim() {
        for claim in [
            Claim::HoldsAtLeast { zatoshi: 50_000_000_000 },
            Claim::HoldsExactly { zatoshi: 1 },
            Claim::ReceivedPayment { txid: [0xcd; 32], zatoshi: 1_200_000_000 },
            Claim::ReceivedAtLeastSince { zatoshi: 5, from_height: 3_000_000 },
        ] {
            let e = sample(claim);
            let f = encode(&e);
            let (d, _) = decode(&f).unwrap();
            assert_eq!(d, e);
            assert_eq!(from_base64url(&to_base64url(&f)).unwrap(), f);
        }
    }

    #[test]
    fn distinct_errors() {
        let f = encode(&sample(Claim::HoldsAtLeast { zatoshi: 12_500_000 }));
        assert_eq!(decode(&f[..20]).unwrap_err(), DecodeError::TooShort);
        let mut bad = f.clone();
        bad[0] = b'X';
        assert_eq!(decode(&bad).unwrap_err(), DecodeError::BadMagic);
        let mut bad = f.clone();
        bad[4] = 2;
        assert_eq!(decode(&bad).unwrap_err(), DecodeError::UnknownVersion(2));
        let mut bad = f.clone();
        bad[100] ^= 1;
        assert_eq!(decode(&bad).unwrap_err(), DecodeError::Checksum);
    }

    fn body_err(e: &Envelope) -> DecodeError {
        decode(&encode(e)).unwrap_err()
    }

    #[test]
    fn varints_are_canonical() {
        let read = |b: &[u8]| Reader { b, o: 0 }.varint();
        assert_eq!(read(&[0x00]), Ok(0));
        assert_eq!(read(&[0x80, 0x01]), Ok(128));
        for padded in [&[0x80, 0x00][..], &[0x81, 0x00], &[0xff, 0x80, 0x00]] {
            assert_eq!(read(padded), Err("non-canonical varint".into()), "{padded:02x?}");
        }
        for n in [0, 1, 127, 128, 300, u32::MAX as u64, u64::MAX] {
            let mut w = Writer(vec![]);
            w.varint(n);
            assert_eq!(read(&w.0), Ok(n));
        }

        // The same proof with its claim tag padded to two bytes is refused, not a second valid file.
        let f = encode(&sample(Claim::HoldsAtLeast { zatoshi: 12_500_000 }));
        let body = [&[0x80u8, 0x00][..], &f[7..f.len() - 32]].concat();
        let padded = [&f[..6], &body[..], &blake2b_256(&body)[..]].concat();
        assert_eq!(decode(&padded).unwrap_err(), DecodeError::Body("non-canonical varint".into()));
    }

    #[test]
    fn lengths_are_bounded_before_narrowing() {
        let mut w = Writer(vec![]);
        w.varint((1 << 32) + 5);
        assert!(Reader { b: &w.0, o: 0 }.len(16, |n| n.to_string()).is_err());
        let mut e = sample(Claim::HoldsAtLeast { zatoshi: 12_500_000 });
        e.evidence.public_inputs = vec![[0; 32]; MAX_PUBLIC_INPUTS + 1];
        assert_eq!(body_err(&e), DecodeError::Body("too many public inputs".into()));
    }

    #[test]
    fn timestamps_are_bounded_and_ordered() {
        let e = sample(Claim::HoldsAtLeast { zatoshi: 12_500_000 });
        let mut late = e.clone();
        late.expires_at = MAX_TIMESTAMP + 1;
        assert_eq!(body_err(&late), DecodeError::Body("timestamp exceeds 2^53 - 1".into()));
        let mut backwards = e.clone();
        backwards.issued_at = e.expires_at + 1;
        assert_eq!(body_err(&backwards), DecodeError::Body("issued after it expires".into()));
        let mut edge = e;
        (edge.issued_at, edge.expires_at) = (MAX_TIMESTAMP, MAX_TIMESTAMP);
        assert_eq!(decode(&encode(&edge)).unwrap().0, edge);
    }

    #[test]
    fn statement_excludes_evidence() {
        let a = sample(Claim::HoldsAtLeast { zatoshi: 12_500_000 });
        let mut b = a.clone();
        b.evidence.proof[0] ^= 1;
        assert_eq!(statement_hash(&a), statement_hash(&b));
        b.expires_at += 1;
        assert_ne!(statement_hash(&a), statement_hash(&b));
    }
}
